### PR TITLE
geom

### PR DESCRIPTION
**Pull request by jschueller**
- Add Mesh.intersect(Mesh) method (2-d only) for UniformOverMesh CDF
- Implement Mesh pretty-print to just print sizes: `Mesh(dimension=1, vertices=4, simplices=3)`
- Improve TruncatedOverMesh/UniformOverMesh computeCDF

### CODE REVIEW

File: lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx [hunk 1]
```diff
@@ -210,31 +211,29 @@ Scalar TruncatedOverMesh::computePDF(const Point & point) const
...
-    Indices simplexIndices(simplexVertices.getSize());
-    simplexIndices.fill();
-    const Mesh simplexMesh(simplexVertices, IndicesCollection(Collection<Indices>(1, simplexIndices)));
-    cdf += integrationAlgorithm.integrate(PDFWrapper(distribution_.getImplementation()->clone()), simplexMesh)[0];
+    try
+    {
+      const Mesh intersectionMesh(mesh_.intersect(IntervalMesher(Indices(getDimension(), 1)).build(intersection)));
+
+      // integrate pdf over simplices of the quadrant/mesh intersection
+      SimplicialCubature integrationAlgorithm;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx:228]: const We should probably add specific keys in ResourceMap as the default number of evaluations of SimplicialCubature is too low for many applications including this one

**Review by regislebrun — APPROVED** [review]
Only cosmetic and minor remarks

**Comment by regislebrun** [comment]
Is there any candidate library to extend the intersection to the 3D case? CGAL? Or a hand-made algorithm, with this algorithm to reject non-ovelaping tetrahedra https://vcg.isti.cnr.it/Publications/2003/GPR03/fast_tetrahedron_tetrahedron_overlap_algorithm.pdf

File: lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx (hunk 1 above)
Comment (jschueller) [reply lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx:228]: we have already SimplicialCubature-DefaultMaximumCallsNumber ?

File: lib/src/Base/Geom/Mesh.cxx [hunk 2]
```diff
@@ -998,6 +1006,92 @@ void Mesh::exportToVTKFile(const String & fileName,
...
+Mesh Mesh::getSubMesh(const Indices & simpliciesIndices) const
+{
+  if (!simpliciesIndices.check(simplices_.getSize()))
+    throw InvalidArgumentException(HERE) << "Simplices indices must be in [0, " << simplices_.getSize() << "[";
+  const UnsignedInteger simplexSize = getDimension() + 1;
+  IndicesCollection simplices(simpliciesIndices.getSize(), simplexSize);
+  for (UnsignedInteger i = 0; i < simpliciesIndices.getSize(); ++ i)
+    for (UnsignedInteger j = 0; j < simplexSize; ++ j)
+      simplices(i, j) = simplices_(simpliciesIndices[i], j);
+  Mesh result(vertices_, simplices);
```
[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Base/Geom/Mesh.cxx]: yes we should

File: lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx (hunk 1 above)
Comment (jschueller) [reply lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx:228]: oh you mean a key to parametrize SimplicialCubature in the context of TruncatedOverMesh
lets reuse TruncatedOverMesh-MaximumIntegrationNodesNumber here

**Comment by jschueller** [comment]
I did not find one unfortunately, also most CGAL stuff is GPL unfortunately.
I will continue looking into 3d intersection anyway.

File: lib/src/Base/Geom/Mesh.cxx [hunk 3]
```diff
@@ -998,6 +1006,92 @@ void Mesh::exportToVTKFile(const String & fileName,
...
+  for (UnsignedInteger i1 = 0; i1 < getSimplicesNumber(); ++ i1)
+  {
+    polygon_t tri1;
+    for(UnsignedInteger j1 = 0; j1 < 4; ++ j1)
+    {
+      // the first vertex is repeated at the end
+      const Point pj1(vertices_[simplices_(i1, j1 % 3)]);
+      boost::geometry::append(tri1.outer(), point_t(pj1[0], pj1[1]));
+    }
+    if (boost::geometry::area(tri1) < 0.0)
```
[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Base/Geom/Mesh.cxx:1035]: yes; todo: also recompression

**Review by josephmure — CHANGES_REQUESTED** [review]

**Pull request merged**